File: Data Structures & Algorithms/house-robber-ii/submission-4.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        def solve(arr):
            memo = [-1] * len(arr)

            def dfs(i):
                if i >= len(arr):
                    return 0
                
                if memo[i] != -1:
                    return memo[i]
                
                memo[i] = max(dfs(i + 1), arr[i] + dfs(i + 2))
                return memo[i]
            
            return dfs(0)
        
        return max(solve(nums[1:]), solve(nums[:-1]))

File: Data Structures & Algorithms/jump-game/submission-4.cpp
class Solution {
private:
    int solve(const vector<int>& nums, vector<int>& memo, int i) {
        if (i == 0) return 1;
        if (memo[i] != -1) return memo[i];

        for (int j = i - 1; j >= 0; j--) {
            if (solve(nums, memo, j) && j + nums[j] >= i) {
                return memo[i] = 1;
            }
        }

        return memo[i] = 0;
    }

public:
    bool canJump(vector<int>& nums) {
        // vector<int> memo(nums.size(), -1);
        // return solve(nums, memo, nums.size() - 1) == 1;

        int max_reach = 0;
        int n = nums.size();

        for (int i = 0; i < n; i++) {
            if (i > max_reach) return false;

            max_reach = max(max_reach, i + nums[i]);

            if (max_reach >= n - 1) return true;
        }

        return true;
    }
};